### DISCUSSION TITLE
Sqlalchemy read Oracle BLOB field is very slow.

### DISCUSSION THREAD

**Discussion by PaleNeutron** [sqlalchemy/sqlalchemy#12014] [Usage Questions] [answered]
sql:

```
SELECT user.emp_id,
       user.email,
       user.name,
       t_empe.empe_portrait
FROM user
JOIN t_empe ON user.emp_id = t_empe.empe_no
```

ddl:

```
CREATE TABLE test_schema.T_EMPE (
	EMPE_NO VARCHAR2(30),
	EMPE_PORTRAIT BLOB,
	UPDDATE_TIME DATE DEFAULT sysdate

);

CREATE TABLE test_schema.user (
	EMP_ID VARCHAR2(20),
	NAME VARCHAR2(200),
	EMAIL VARCHAR2(80),
);
```

Singe `EMPE_PORTRAIT` is less than 100kb and user table is less than 500 rows.

This sql could get result in less than 1s with dbeaver or oracledb(both thin and thick mode):

![image](https://github.com/user-attachments/assets/b9a9a010-8a5c-4539-b79a-606893125253)

but `8m52s` with sqlalchemy 2.0.32:

![image](https://github.com/user-attachments/assets/c434c558-e656-4dac-9762-dd105c23ccb9)

**Comment by PaleNeutron** [comment]
I found the problem, if I append `oracledb.defaults.fetch_lobs = False` to my test code, `oracledb` will also takes 8min.
.

**Comment by CaselIT** [comment, marked as answer]
Hi,

Is this the option you are looking for? https://docs.sqlalchemy.org/en/20/dialects/oracle.html#lob-datatypes
Otherwise I would need to understand what oracledb does differently here. Does it read the data in the lob column or does it return to you a stream that you need to manually read?

**Comment by PaleNeutron** [comment]
@CaselIT  BTW, I found query rows contains blob with `pymysql` or `mysqlclient` is much slower (10s vs 0.5s) than GUI Client such as dbeaver or MySQL workbench.

What makes the difference?

**Reply by CaselIT** [reply]
Does using the dialect directly like in your example that used oracledb, has the same time difference?

Overall I'm not sure if sqlalchemy should do something different here.

**Reply by PaleNeutron** [reply]
It's nothing with sqlalchemy, I tested it with pure `pymysql` and `mysqlclient`, just very curious about if blob cause slow query in python is a widely known issue.

The table ddl is very similar to the oracle version above.

**Reply by CaselIT** [reply]
I think that unless the driver does something particular, lob will require that for each blob in each row you call `.read()` or equivalent, so another call to the db is needed.

At least that's what sqlalchemy does by default when processing the lobs in the rows.
